### PR TITLE
Feature: Made it possible to post a log from the overview screen

### PR DESCRIPTION
**Pull request by john-ghatas**

### CODE REVIEW

File: lib/domain/dtos/CreateLogDto.js [hunk 1]
```diff
@@ -16,14 +16,19 @@ const Joi = require('@hapi/joi');
 const BodyDto = Joi.object({
     title: Joi.string()
         .required()
-        .min(3),
+        .min(3)
+        .max(140),
```
Comment (mvegter) [inline lib/domain/dtos/CreateLogDto.js:20]: I would expect this change to also be reflected in the openapi specification and an e2e test.

File: lib/public/Model.js [hunk 2]
```diff
@@ -61,6 +61,8 @@ export default class Model extends Observable {
             case 'entry':
                 this.logs.fetchOneLog(this.router.params.id);
                 break;
+            case 'create':
```
Comment (mvegter) [inline lib/public/Model.js:64]: Using `create` is too generic as in the (short-term) future more create pages will be required.

File: lib/public/app.css [hunk 3]
```diff
@@ -1,3 +1,5 @@
+.btn-right {float: right }
```
Comment (mvegter) [inline lib/public/app.css:1]: Add an extra space inside the object

File: lib/public/views/Logs/Logs.js [hunk 4]
```diff
@@ -50,6 +53,70 @@ export default class Overview extends Observable {
...
+                title: this.getTitle(),
+                text: this.getText(),
+            }),
+        });
+
+        const result = await response.json();
+
+        if (!(result === undefined)){
+            this.error = false;
+            this.model.router.go('/');
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:77]: Is the expected behaviour to go to the main page or to the detail view (where an user can add tags / attachemtens / ... )?

**Comment by john-ghatas** [comment]
> The last commit should be squashed with the first commit. Also remove the emoji from the commit message.

Done

File: lib/public/views/Logs/Create/index.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    const text = model.logs.getText();
+    const disabled = !(title && text);
+    const error = model.logs.hasErrorOccured();
+
+    return h('div#create-log', [
+        h('', [
+            h('h3.black.line-break: auto', 'Title of the log'),
+            h('input#title', {
+                placeholder: 'Enter the title of the log entry...',
+                maxlength: 140,
```
Comment (mvegter) [inline lib/public/views/Logs/Create/index.js:31]: You should also include the minLength property.

File: lib/public/views/Logs/Logs.js [hunk 6]
```diff
@@ -50,6 +53,70 @@ export default class Overview extends Observable {
...
+                Accept: 'application/json',
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({
+                title: this.getTitle(),
+                text: this.getText(),
+            }),
+        });
+
+        const result = await response.json();
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:73]: The response is valid when it has the data property containing the created log (useful for a redirect). The response (or request) is invalid if it has the error property, this should allow for better error reporting.

File: lib/public/views/Logs/Logs.js [hunk 7]
```diff
@@ -50,6 +53,70 @@ export default class Overview extends Observable {
...
+
+        if (!(result === undefined)){
```
Comment (MauritsioRK) [inline lib/public/views/Logs/Logs.js:75]: A slightly shorter way to write this: `result !== undefined`

File: test/public/logs/overview.test.js [hunk 8]
```diff
@@ -33,6 +33,27 @@ async function findRowById(table, page) {
...
+/**
+ * //  * A function to check if a table row contains the title that has been posted
+ * //  * @param {Object} parent The parent element to be examined
+ * //  * @param {Object} page The page puppeteer is currently running
+ * //  * @param {String} string The string that should be in the log entry
+ * //  * @returns {Boolean} Returns wether the title has been posted correctly
+ * //
+ */
+/*
+ * Async function checkMatchingString(parent, page, string) {
```
Comment (MauritsioRK) [inline test/public/logs/overview.test.js:45]: Should you keep this kind of dead code?

File: lib/public/views/Logs/Create/index.js (hunk 5 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Create/index.js:32]: The minlength property has ben added to the input, this has also been incorporated with the button visibility check

File: lib/public/views/Logs/Logs.js (hunk 6 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js:74]: The error handling has been improved as per this comment.

File: lib/public/views/Logs/Logs.js (hunk 4 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js]: This feature has been added

File: lib/domain/dtos/CreateLogDto.js (hunk 1 above)
Comment (MauritsioRK) [reply lib/domain/dtos/CreateLogDto.js:20]: I've helped John to create an e2e test and updated the openapi spec.

File: lib/public/Model.js (hunk 2 above)
Comment (john-ghatas) [reply lib/public/Model.js]: Renamed it to create-log-entry

File: lib/public/app.css (hunk 3 above)
Comment (john-ghatas) [reply lib/public/app.css]: Done

**Comment by john-ghatas** [comment]
Squashed the revirw commit. Will be looking into the redirect in a bit

**Comment by john-ghatas** [comment]
The issue is fixed by going to the root page then routing to the detail view.

File: lib/public/views/Logs/Logs.js [hunk 9]
```diff
@@ -50,14 +55,89 @@ export default class Overview extends Observable {
...
+        const result = await response.json();
+
+        if (result.data) {
+            const logId = await result.data.id;
+            this.fetchedFromCreation = true;
+            await this.model.router.go(`/?page=entry&id=${logId}`);
+            this.error = false;
+        } else {
+            this.error = true;
+            this.errorMessage = result.errors[0].detail;
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:84]: It is an array containing one or more errors, only showing the first is not that usefull I think.

File: lib/public/views/Logs/Logs.js [hunk 10]
```diff
@@ -143,7 +223,9 @@ export default class Overview extends Observable {
      * @return {Boolean} Returns the status of the existence of a tag in the data entry
      */
     hasExistingTag(entry) {
-        return this.filterCriteria.filter((tag) => entry.tags.filter(({ text }) => tag === text).length > 0).length > 0;
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:146]: Not sure why this change is needed.

File: lib/public/views/Logs/Create/index.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                rows: 5,
+                oninput: (e) => model.logs.setText(e.target.value),
+            }, text),
+        ]),
+        h('button.btn.btn-success.m#send', {
+            onclick: () => model.logs.createLog(),
+            disabled,
+        }, 'Create'),
+        error &&
+            h('span.danger.mh2#error', errorMessage || 'There was a problem with posting the log, try again later'),
```
Comment (mvegter) [inline lib/public/views/Logs/Create/index.js:52]: I would expect the errors to be the item at the top of the screen, also have a look at [Bootstrap alerts](https://getbootstrap.com/docs/4.0/components/alerts/#dismissing).

File: lib/public/views/Logs/Logs.js [hunk 12]
```diff
@@ -50,14 +55,89 @@ export default class Overview extends Observable {
...
+        return this.text;
+    }
+
     /**
      * Retrieve a specified log from the API
      * @param {Number} id The ID of the log to be found
      * @returns {undefined} Injects the data object with the response data
      */
     async fetchOneLog(id) {
-        // Do not call this endpoint again if we already have all the logs
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:59]: Comment indent should be aligned with the code.

File: lib/public/views/Logs/Logs.js (hunk 9 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js:84]: Done, mapping through the errors and showing them on top now will be fixed in the next commit

File: lib/public/views/Logs/Create/index.js (hunk 11 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Create/index.js:52]: This one as well

File: lib/public/views/Logs/Logs.js (hunk 10 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js]: Fixed

File: lib/public/views/Logs/Logs.js (hunk 12 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js:59]: Locally the comment shows the right indentation, don't know why it shows up like this on github

Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js:59]: image

File: lib/public/views/Logs/Create/index.js (hunk 11 above)
Comment (mvegter) [reply lib/public/views/Logs/Create/index.js]: image

Errors are being chained while I would expect them to be on a newline

File: lib/public/views/Logs/Logs.js [hunk 13]
```diff
@@ -50,14 +55,89 @@ export default class Overview extends Observable {
...
+            body: JSON.stringify({
+                title: this.getTitle(),
+                text: this.getText(),
+            }),
+        });
+
+        const result = await response.json();
+
+        if (result.data) {
+            const logId = await result.data.id;
```
Comment (mvegter) [inline lib/public/views/Logs/Logs.js:78]: Are property getters async?

Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js:78]: This was residue code from debugging the automatic redirect

File: lib/public/views/Logs/Create/index.js (hunk 11 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Create/index.js]: The lines are breaking properly now

File: lib/public/views/Logs/Logs.js (hunk 13 above)
Comment (john-ghatas) [reply lib/public/views/Logs/Logs.js]: Fixed

**Comment by john-ghatas** [comment]
The issues have been resolved in the latest commit
